analyzers/src/SonarAnalyzer.Common/Rules/ParametersCorrectOrderBase.cs

```diff
@@ -20,139 +20,57 @@
-                : base(identifierName, argumentSyntax)
-            {
-                DeclaredName = declaredName;
-            }
-
-            public string DeclaredName { get; }
-        }
+                    if (!c.IsRedundantPrimaryConstructorBaseTypeContext()
+                        && Language.Syntax.ArgumentList(c.Node) is { Count: >= 2 } argumentList // there must be at least two arguments to be able to swap
+                        && Language.MethodParameterLookup(c.Node, c.SemanticModel) is var methodParameterLookup)
+                    {
+                        foreach (var argument in argumentList)
+                        {
+                            // Example void M(int x, int y) <- p_x and p_y are the parameter
+                            // M(y, x); <- a_y and a_x are the arguments
+                            if (methodParameterLookup.TryGetSymbol(argument, out var parameterSymbol) // argument = a_x and parameterSymbol = p_y
+                                && parameterSymbol is { IsParams: false }
+                                && ArgumentName(argument) is { } argumentName // "x"
+                                && !MatchingNames(parameterSymbol, argumentName)  // "x" != "y"
+                                && Language.Syntax.NodeExpression(argument) is { } argumentExpression
+                                && c.Context.SemanticModel.GetTypeInfo(argumentExpression).ConvertedType is { } argumentType
+                                // is there another parameter that seems to be a better fit (name and type match): p_x
+                                && methodParameterLookup.MethodSymbol.Parameters.FirstOrDefault(p => MatchingNames(p, argumentName) && argumentType.DerivesOrImplements(p.Type)) is { IsParams: false }
+                                // is there an argument that matches the parameter p_y by name: a_y
+                                && Language.Syntax.ArgumentList(c.Node).FirstOrDefault(x => MatchingNames(parameterSymbol, ArgumentName(x))) is { })
+                            {
+                                var secondaryLocations = methodParameterLookup.MethodSymbol.DeclaringSyntaxReferences
+                                    .Select(s => Language.Syntax.NodeIdentifier(s.GetSyntax())?.GetLocation())
+                                    .WhereNotNull();
+                                c.ReportIssue(Diagnostic.Create(SupportedDiagnostics[0], PrimaryLocation(c.Node), secondaryLocations, properties: null, methodParameterLookup.MethodSymbol.Name));
+                                return;
+                            }
+                        }
+                    }
+                }, InvocationKinds);
+
+        protected virtual Location PrimaryLocation(SyntaxNode node)
+            => node.GetLocation();
+
+        private static bool MatchingNames(IParameterSymbol parameter, string argumentName) =>
+            parameter.Name == argumentName;
```

Comment:
`ArgumentName()` already returns names for all kinds of expressions:
* `M(A.B)` -> B
* `M(A?.B)` -> B
* `M(A())`-> A
* `M(A)` -> A (this might be unsupported at the moment)

and more could be added when this is fixed https://github.com/SonarSource/sonar-dotnet/issues/8241

Given that the compassion should be replaced by `StringComparison.OrdinalIgnoreCase` for all languages.